text/0000-fn-shorthand.md

```diff
@@ -0,0 +1,470 @@
+contained units that respect the *[single responsibility principle (SRP)][SRP]*,
+*[separation of concerns (SoC)][SoC]* or *modularity* in general.
+
+## Embracing the expression oriented nature
+
+[expr_oriented]: https://doc.rust-lang.org/book/first-edition/glossary.html#expression-oriented-language
+
+As the book tells us, [Rust is an expression-oriented language][expr_oriented].
+Therefore it is somewhat peculiar that one of the most important parts
+of the language, the function definition body, is not an expression as
+in other expression oriented languages (see the [prior-art]).
+In this RFC, we make the function body into an expression.
+
+## A more syntactically unified feeling
+
+Finally, a rather minor motivation, which still bears mentioning, is
+that by permitting `fn foo(...) -> R = expr;` as a syntax, we bring `const`
+and `static` items closer to `fn` items syntactically which gives a smoother
+experience overall. In other words, we can now write:
+
+```rust
+static WIBBLE: Wobble = make_wobble();
+
+const FOO: Bar = make_bar();
+
+fn baz() -> Quux = make_quux();
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Simply put, this RFC allows you to, in addition to the existing syntax,
+define the body of a function, whether the function is free or associated,
+with the following shorthand syntax:
+
+```rust
+fn foo(arguments...) -> ReturnType = expression;
+
+// If the return type is () you can omit `-> ReturnType` as usual.
+```
+
+For example, we may write:
+
+```rust
+pub fn self_ty() -> syn::Type = parse_quote!(Self);
+
+impl Params {
+    pub fn empty() -> Self = Params(Vec::new());
+
+    pub fn len(&self) -> usize = self.0.len();
+}
+
+impl UseTracker {
+    pub fn no_track(&mut self) = self.track = false;
+
+    pub fn consume(self) -> syn::Generics = self.generics;
+
+    fn use_type(&mut self, ty: syn::Type) = self.where_types.insert(ty);
+}
+```
+
+This is particularly useful when your function immediately matches on some
+argument such as with:
+
+```rust
+use syn::Fields;
+
+pub fn fields_to_vec(fields: syn::Fields) -> Vec<syn::Field> = match fields {
+    Fields::Named(fields) => fields.named.into_iter().collect(),
+    Fields::Unnamed(fields) => fields.unnamed.into_iter().collect(),
+    Fields::Unit => vec![]
+};
+```
+
+The RFC does not make any distinction between `const fn`s, visibilities,
+`unsafe fn`, `async fn`, `extern` modifiers and so on. That is, wherever
+a function definition can currently have a body, then the shorthand syntax
+as proposed here may be used.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The grammar of `fn` item definitions is amended such that you may also
+use the syntax `= $expr;` for the body of the function instead of
+`{ $stmt* $tail_expr }`. Note that `= $expr;` comes after a `where` clause
+if any such clause is present. Examples include:
```

Comment:
Could you show an example which a `where` clause is present?

```rust
fn size_of_ptr<T>() -> usize 
where 
    T: ?Sized,
= mem::size_of::<*const T>();
```

Reply:
The formatting you used seems consistent with current rustfmt formatting for `type` aliases tho so it seems right...

Reply:
...added your example. :)

Reply:
The rustfmt of the braced version is 

```rust
fn size_of_ptr<T>() -> usize
where
    T: ?Sized,
{
    mem::size_of::<*const T>()
}
```

I don't think the first 3 lines up to `?Sized,` should be changed in the equal-sign version.

Reply:
Yeah makes sense; I was mostly unsure whether to indent the `=` or not ;)